internal/build.go

```diff
@@ -45,66 +44,62 @@ func decodeInto(ctx context.Context, dest *map[string]interface{}, in io.Reader)
 	return nil
 }
 
-// Build builds and validates a GOBL envelope from the opts.
+// Build builds and validates GOBL data from build options, and transparently
+// wraps a document in an envelope if needed.
 func Build(ctx context.Context, opts *BuildOptions) (*gobl.Envelope, error) {
-	encoded, err := prepareIntermediate(ctx, opts, docInEnvelopeSchemaData)
+	env, err := parseBuildData(ctx, opts)
 	if err != nil {
 		return nil, err
 	}
 
-	// Prepare an empty envelope as we assume the consumer is providing one already.
-	env := new(gobl.Envelope)
-	if err := json.Unmarshal(encoded, env); err != nil {
-		return nil, echo.NewHTTPError(http.StatusBadRequest, err.Error())
-	}
-	if err := finalizeEnvelope(ctx, env, opts); err != nil {
-		return nil, err
+	// Signed documents should be regarded as immutable.
+	// Attempting to build an already signed document returns an error.
+	if len(env.Signatures) > 0 {
+		return nil, echo.NewHTTPError(http.StatusConflict, "document has already been signed")
 	}
-	return env, nil
-}
 
-// Envelop assumes the incoming BuildOptions define the contents of a document
-// payload and we need to prepare the envelope around it.
-func Envelop(ctx context.Context, opts *BuildOptions) (*gobl.Envelope, error) {
-	encoded, err := prepareIntermediate(ctx, opts, docSchemaData)
-	if err != nil {
-		return nil, err
+	if err := env.Calculate(); err != nil {
+		return nil, echo.NewHTTPError(http.StatusUnprocessableEntity, err.Error())
 	}
 
-	// Prepare a new envelope as the intention is to insert the encoded data
-	env := gobl.NewEnvelope()
-	if err = json.Unmarshal(encoded, env.Document); err != nil {
-		return nil, echo.NewHTTPError(http.StatusBadRequest, err.Error())
+	if err := env.Validate(); err != nil {
+		return nil, echo.NewHTTPError(http.StatusUnprocessableEntity, err.Error())
 	}
 
-	if err := finalizeEnvelope(ctx, env, opts); err != nil {
-		return nil, err
-	}
 	return env, nil
 }
 
-type schemaDataCB func(schema.ID) map[string]interface{}
+func parseBuildData(ctx context.Context, opts *BuildOptions) (*gobl.Envelope, error) {
```

Comment:
If we decide to rename `BuildOptions` to something that makes sense for the commands that need to parse incoming data, we can rename this func as well.